Additional source control functionality

austincondiff opened the pull request:
### Description

We are improving upon existing source control functionality by doing the following...

- Added push functionality and sheet. 
- Added pull functionality and sheet. 
- Added manual fetch functionality and fetching sheet.
- Reorganized source control files in project
    - Renamed Features/Git to Features/SourceControl
    - Moved sheet views into it from the Source Control Navigator
    - Moved models out of Client
- Moved the sync view above commit form in source control changes navigator. 
- Refreshing remotes cache after adding a remote. 
- Moved push and pull state variables into the source control manager. 
- Added a confirm branch switch sheet.
- Stashing changes if changes exist when attempting to pull or switch branches. 
- Setting remote for push and pull after adding a remote.
- Added warning alerts if uncommitted changes are not present
- It is now possible to switch to a remote branch and create a new branch from a remote branch. 
- Disabling rename branch for remote branches.


### Testing

Note to reviewers... Please pull down and test this in addition to reviewing the code. Once you have this pulled down...

In a terminal run...
```
mkdir test.git
cd test.git
git init --bare
```
Then add the remote in CodeEdit by 
1. going to the Source Control Navigator
2. If not already click Initiate to start a git repo
3. go to the Repositories tab
4. Right click Remotes and click Add existing remote...
5. Give the remote a name like "test" (the first remote will be "origin") and enter the path to test.git
7. You can now make changes, commit, and push via the Changes tab in the Source Control Navigator
6. You can set up additional projects by repeating steps 1-5 so you can make changes elsewhere and test pull functionality.
7. You can create additional remotes as well to test managing a project with multiple remotes, pushing and pulling from them, etc.

### Checklist

<!--- Add things that are not yet implemented above -->

- [x] Add push logic
- [x] Add pull logic
- [x] Add push sheet
- [x] Add pull sheet
- [x] Add centralized sheets view
- [x] Rename Features/Git to Features/SourceControl
- [x] Move sheet views into Features/SourceControl
- [x] Add "Source Control" menu
- [x] Add remote from push and pull sheets
- [x] Set new remote in push or pull sheet after one is added
- [x] Stash changes if attempting to pull or switch branches and changes exist
- [x] Add fetch to the menu and display fetching sheet while fetching
- [x] Add confirm branch switch sheet
- [x] Add warning alerts if uncommitted changes are not present
---
- [x] I read and understood the [contributing guide](https://github.com/CodeEditApp/CodeEdit/blob/main/CONTRIBUTING.md) as well as the [code of conduct](https://github.com/CodeEditApp/CodeEdit/blob/main/CODE_OF_CONDUCT.md)
- [x] The issues this PR addresses are related to each other
- [x] My changes generate no new warnings
- [x] My code builds and runs on my machine
- [x] My changes are all related to the related issue above
- [x] I documented my code

### Screenshots

<img width="598" alt="image" src="https://github.com/CodeEditApp/CodeEdit/assets/806104/e599cb12-0772-4a6e-abca-b5f9af7fa74a">

<img width="598" alt="image" src="https://github.com/CodeEditApp/CodeEdit/assets/806104/eca3d546-d38c-4787-89ed-9930220fd4e4">

<img width="598" alt="image" src="https://github.com/CodeEditApp/CodeEdit/assets/806104/20a5c002-5fff-4920-8211-8333fce56f6b">

<img width="598" alt="image" src="https://github.com/CodeEditApp/CodeEdit/assets/806104/6e80c67b-5356-4cf7-82b7-1ad95a13e05b">

<img width="342" alt="image" src="https://github.com/CodeEditApp/CodeEdit/assets/806104/0e55f5f8-f71b-4e30-bc05-d80a5796360e">

<img width="712" alt="image" src="https://github.com/CodeEditApp/CodeEdit/assets/806104/00c61863-5cb4-409b-b2d6-40098db5c57e">

### Todos discovered

- Reload source editor contents when file is changed externally
- Handle source control errors better and show necessary errors to user
- Handle rebase and merge conflict flow in UI
- ~~Display loading indicator when pushing or pulling~~ Be able to cancel push or pull while in progress
- Display loading indicator when pulling from stash sheet
- Write tests for source control features

austincondiff commented:
I am struggling to get the "Source Control" menu working, specifically trying to gain access to the focused windows `SourceControlManager`. The commands view can't find it so we are not showing those items as a result. Anyone know why? I tried placing `.focusedObject(sourceControlManager)` at the end of `WorkspaceView` but that wasn't working.

FastestMolasses commented on CodeEdit/WorkspaceView.swift:
```diff
@@ -80,6 +79,17 @@ struct WorkspaceView: View {
                         }
                     }
+                    .onChange(of: focusedEditor) { newValue in
+                        /// update active tab group only if the new one is not the same with it.
```
Capitalize first character in comment

FastestMolasses reviewed (commented):
LGTM, just small changes

austincondiff commented on CodeEdit/Features/SourceControl/SourceControlManager+GitClient.swift:
```diff
@@ -251,6 +124,94 @@ final class SourceControlManager: ObservableObject {
+    }
+
+    /// Refresh all changed files and refresh status in file manager
+    func refreshAllChangedFiles() async {
```
Not 100% sure to be honest. I didn't touch this code as part of this PR. I just split up SourceControlManager to fix a SwiftLint error.

FastestMolasses replied:
Ill make a note of it and we can take care of it in another PR

matthijseikelenboom commented:
Looks good, but no tests are present

austincondiff commented:
@matthijseikelenboom We need a separate issue to add tests to not just what I did in this PR but everything source control related.

nanashili commented on CodeEdit/Features/SourceControl/Client/GitClient+Pull.swift:
```diff
@@ -0,0 +1,25 @@
+            command += " \(remote) \(branch)"
+        }
+
+        if rebase == true {
```
Why are you checking if it's explicitly true?

austincondiff replied:
Because it might be either `nil` or `false`.

nanashili replied:
It never needs to be nil only false or true, i would opt for always returning false instead of true

austincondiff replied:
Done

nanashili commented on CodeEdit/Features/SourceControl/Client/GitClient+Pull.swift:
```diff
@@ -0,0 +1,25 @@
+
+extension GitClient {
+    /// Pull changes from remote
+    func pullFromRemote(remote: String? = nil, branch: String? = nil, rebase: Bool? = nil) async throws {
```
Why are the params optional, these would almost always be required.

austincondiff replied:
Users can do a `git pull` and it will pull their current branch without needing to specify. I am keeping close to how git works in this regard. Probably not extremely necessary.

nanashili replied:
Using just git pull is not very effective when it comes to big projects. I would perhaps rethink how this is managed. Because if something goes wrong with git users won't be too happy

nanashili commented on CodeEdit/Features/SourceControl/Client/GitClient+Pull.swift:
```diff
@@ -0,0 +1,25 @@
+    func pullFromRemote(remote: String? = nil, branch: String? = nil, rebase: Bool? = nil) async throws {
+        var command = "pull"
+
+        if let remote = remote, let branch = branch {
```
This can be done better in my opinion:

```swift
func pullFromRemote(remote: String = "origin",
                    branch: String = "main", 
                    rebase: Bool = false) async throws {
    var command = "pull"

    if remote == "" {
        print("Remote needs to be specified")
        return
    }

    if branch == "" {
        print("Branch needs to specified")
        return
    }

    command += " \(remote) \(branch)"

    if rebase {
        command += " --rebase"
    }

    _ = try await self.run(command)
}
```

austincondiff replied:
You are making the assumption that their default branch is "main".

nanashili replied:
> You are making the assumption that their default branch is "main".

Nope, just an example, because you would always need to pull the origin and branch regardless

matthijseikelenboom replied:
I think you'd want to default to nothing, in case of remote name and branch name.

The rebase value should indeed be non-nullable. But, since there is also the option to use --merge or --ff-only. might want to make it an inner private enum, instead of a boolean.

But if you'd follow clean code, you want four functions:
`pullFromRemoteWithRebase`
`pullFromRemoteWithMerge`
`pullFromRemoteWithFastForward`
`pullFromRemoteWithGlobalConfig`
Or something like that

nanashili replied:
I would suggest making that one function instead of multiple, you would end up with repetitive code if you make multiple functions

activcoding commented:
Great work! One thing that's bothering me is that I don't see the push button until I commit all my changes. Once everything is committed, the button appears.
1. No new commits
<img width="240" alt="Screenshot 2024-07-05 at 1 25 07 PM" src="https://github.com/CodeEditApp/CodeEdit/assets/83090745/14e51824-1003-45c1-bfa7-4e401516bbfa">

2. One new commit
<img width="242" alt="Screenshot 2024-07-05 at 1 25 20 PM" src="https://github.com/CodeEditApp/CodeEdit/assets/83090745/7c95dff3-5815-4fed-bdd6-a65df23c768b">

3. Two new commits
<img width="242" alt="Screenshot 2024-07-05 at 1 25 29 PM" src="https://github.com/CodeEditApp/CodeEdit/assets/83090745/4d2467c5-9bad-4a8d-968c-8175a57a228d">

activcoding reviewed (changes requested):
Amazing work! Source control in CE is really shaping up. I just have a few small comments 😁

matthijseikelenboom commented on CodeEdit/Features/SourceControl/Client/GitClient+Pull.swift:
```diff
@@ -0,0 +1,25 @@
+
+extension GitClient {
+    /// Pull changes from remote
+    func pullFromRemote(remote: String? = nil, branch: String? = nil, rebase: Bool? = nil) async throws {
```
I'd still advocate for separate functions

https://martinfowler.com/bliki/FlagArgument.html
https://luzkan.github.io/smells/flag-argument

activcoding replied:
Great article! Splitting up the function would definitely make the code more readable, though I'm not sure if it applies here. tho if we decide to split the method, I would prefer to keep it private as shown in the article.

austincondiff replied:
I am not going to make any changes to this function as part of this PR.

austincondiff commented on CodeEdit/Features/SourceControl/Client/GitClient+Push.swift:
```diff
@@ -9,10 +9,13 @@ import Foundation
 extension GitClient {
     /// Push changes to remote
-    func pushToRemote(upstream: String? = nil) async throws {
+    func pushToRemote(remote: String? = nil, branch: String? = nil, setUpstream: Bool? = false ) async throws {
```
I was trying to keep the usage clean by not requiring that every property needs to be defined for each usage.

activcoding replied:
I actually mean that there's no reason to use an optional bool instead of a regular bool, since you're defaulting it to false. This means there is no scenario where someone would pass nil.

austincondiff replied:
Done

austincondiff commented on CodeEdit/Features/SourceControl/SourceControlManager+GitClient.swift:
```diff
@@ -1,146 +1,19 @@
-        if updatedStatusFor.isEmpty {
-            return
+extension SourceControlManager {
+    /// Validate repository
```
That was there before, just copied it over. I'd rather not touch it as part of this PR

austincondiff commented on CodeEdit/Features/SourceControl/Views/SourceControlPushView.swift:
```diff
@@ -0,0 +1,76 @@
+                await scm.showAlertForError(title: "Failed to push", error: error)
+            }
+        }
+    }
```
This is how other similar views are. Is this a convention we've agreed upon?

activcoding replied:
That's how it's mostly done in the Swift community.

austincondiff replied:
Done.

austincondiff commented on CodeEdit/Features/SourceControl/SourceControlManager+GitClient.swift:
```diff
@@ -251,6 +124,94 @@ final class SourceControlManager: ObservableObject {
+            }
+
+            // TODO:  Profile
+            let changedFiles = Array(fileDictionary.values.sorted())
```
I believe it is necessary because otherwise it would show random every time in the UI

austincondiff commented on CodeEdit/Features/SourceControl/SourceControlManager.swift:
```diff
@@ -0,0 +1,115 @@
+//  SourceControlModel.swift
+//  CodeEdit
+//
+//  Created by Nanashi Li on 2022/05/20.
```
I didn't originally add this. Just reorganized into a different folder.

austincondiff commented on CodeEdit/Features/SourceControl/SourceControlManager.swift:
```diff
@@ -0,0 +1,115 @@
+import Foundation
+import AppKit
+
+/// This model handle the fetching and adding of changes etc... 
```
I didn't touch this part of the file.

austincondiff replied:
I added a more detailed comment

austincondiff commented on CodeEdit/Features/SourceControl/Client/GitClient+Pull.swift:
```diff
@@ -0,0 +1,25 @@
+
+extension GitClient {
+    /// Pull changes from remote
+    func pullFromRemote(remote: String? = nil, branch: String? = nil, rebase: Bool? = nil) async throws {
```
Rebase now defaults to `false`.

austincondiff commented:
> Great work! One thing that's bothering me is that I don't see the push button until I commit all my changes. Once everything is committed, the button appears.
> 
> 1. No new commits
> 
> <img alt="Screenshot 2024-07-05 at 1 25 07 PM" width="240" src="https://private-user-images.githubusercontent.com/83090745/346109408-14e51824-1003-45c1-bfa7-4e401516bbfa.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjA1NjI5MDEsIm5iZiI6MTcyMDU2MjYwMSwicGF0aCI6Ii84MzA5MDc0NS8zNDYxMDk0MDgtMTRlNTE4MjQtMTAwMy00NWMxLWJmYTctNGU0MDE1MTZiYmZhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MDklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzA5VDIyMDMyMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTJjZGU4M2E3NmVhNjlmNjAzNGQ0ZDE1MjYwNjE3MzUxY2Y3ZmVjNmUzNWI0MGZhNTA0NjQyM2FkMTViYjY4NzMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.9jcgLKPfN-Fuk4Ql2xa6sOnG7veHu2Rp6BdHcMSd_9o">
> 2. One new commit
> 
> <img alt="Screenshot 2024-07-05 at 1 25 20 PM" width="242" src="https://private-user-images.githubusercontent.com/83090745/346109407-7c95dff3-5815-4fed-bdd6-a65df23c768b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjA1NjI5MDEsIm5iZiI6MTcyMDU2MjYwMSwicGF0aCI6Ii84MzA5MDc0NS8zNDYxMDk0MDctN2M5NWRmZjMtNTgxNS00ZmVkLWJkZDYtYTY1ZGYyM2M3NjhiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MDklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzA5VDIyMDMyMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWZhNjkyNDRkYTU5ODVlOGU3NzczZWU0ZTkyZDkyODYyMGI2MDYxNTM1M2ZkMzgxZWM2ZmM0NzllN2Y2ZmFhM2ImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.ThomuswuuZEmY0cI2k8Mz45khJBDuyTbROtQo1BKpng">
> 3. Two new commits
> 
> <img alt="Screenshot 2024-07-05 at 1 25 29 PM" width="242" src="https://private-user-images.githubusercontent.com/83090745/346109400-4d2467c5-9bad-4a8d-968c-8175a57a228d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjA1NjI5MDEsIm5iZiI6MTcyMDU2MjYwMSwicGF0aCI6Ii84MzA5MDc0NS8zNDYxMDk0MDAtNGQyNDY3YzUtOWJhZC00YThkLTk2OGMtODE3NWE1N2EyMjhkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MDklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzA5VDIyMDMyMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTU3MWQ1MmY4NzQxNjAyNDE4NGM2OWE1Y2EzZmY2ZTVkYmI4NmM2YTA2MTE0MjYzMWI0MjY5ZTdlYWQ4N2FkODMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.97WcbW1M9tRfLt0z7_itLtpxZGmhyjfN5IeE4qPt43c">

@activcoding The sync bar now shows even if you have changes.

thecoolwinter commented on CodeEdit/Features/SourceControl/Views/SourceControlAddExistingRemoteView.swift:
```diff
@@ -71,4 +66,22 @@ struct SourceControlAddRemoteView: View {
+    func submit() {
+        Task {
+            do {
+                try await sourceControlManager.addRemote(name: name, location: location)
```
Might be a good idea to add a quick `.trimmingWhitespace` on the name and location here. I found that both text fields allowed for pasting newlines and didn't easily show that there was one at the end. 

Actually this is maybe an input validation problem rather than adding it here. I'll leave that decision to you.

thecoolwinter reviewed (changes requested):
Should this state be possible? This is an empty git repository with no remotes. I think it shouldn't show the no upstream branch if there's no tracking remote.

![Screenshot 2024-07-11 at 5 17 53 PM](https://github.com/user-attachments/assets/f7662f15-b4c1-432a-a6a2-496e99a7a93b)

Not quite done with the full review, I'll finish it when I have some time.

thecoolwinter commented on CodeEdit/Features/SourceControl/Views/SourceControlNavigatorNewBranchView.swift:
I know this wasn't changed in this PR but it's related. The labels on the fields here aren't great. Could we rename them to "From Branch" and "New Branch Name"

austincondiff replied:
I personally like them as is, especially considering this is what Xcode uses...

<img width="549" alt="image" src="https://github.com/user-attachments/assets/486f78c6-ae06-47e6-80b6-9239e7378415">

thecoolwinter commented on CodeEdit/Features/NavigatorArea/SourceControlNavigator/Repository/Views/SourceControlNavigatorRepositoryView+outlineGroupData.swift:
```diff
@@ -60,6 +60,15 @@ extension SourceControlNavigatorRepositoryView {
                             label: remote.name,
                             symbolImage: "vault",
                             imageColor: .teal,
+                            children: remote.branches?.map { branch in
```
One small side effect of this is that I can use the "New Branch From" menu item on these, but the flow doesn't seem to actually work on remote branches. Is there a way to disable that menu item for these for now until that feature can be implemented?

![Screenshot 2024-07-11 at 10 05 10 PM](https://github.com/user-attachments/assets/1a07e654-3726-4a8f-9f4c-83c16552666c)

austincondiff replied:
> It looks like theres a way to make the default selected branch in this menu be incorrect. This was after making a new branch, switching to it, and pressing the push button. It should default to the selected branch but it seems the operationBranch wasn't updated correctly from the repository outline view.

This seems to work for me. Can you reproduce this with the latest?

austincondiff replied:
I implemented it. You can now switch to a remote branch and create a new branch from a remote branch. I also disabled rename branch for remote branches.

thecoolwinter reviewed (changes requested):
Continuing my last review.

It looks like theres a way to make the default selected branch in this menu be incorrect. This was after making a new branch, switching to it, and pressing the push button. It should default to the selected branch but it seems the `operationBranch` wasn't updated correctly from the repository outline view.

![Screenshot 2024-07-11 at 10 06 32 PM](https://github.com/user-attachments/assets/e2d2a3fa-5606-434f-b211-4217609aaa67)

Also, I seem to be able to break the push mechanism on my machine:
https://github.com/user-attachments/assets/a0c9cc34-03eb-426c-8c6a-1295973e09ff
For more details:
- I created a branch, switched to it, pushed it to the remote, and switched back w/o closing the window
- After switching back, the push mechanism was broken.

Last thing: Could you provide more details on how you tested pulling changes? I can't figure how you did it with a local git repository.

activcoding commented on CodeEdit/Features/SourceControl/Models/GitRemote.swift:
```diff
@@ -11,4 +11,5 @@ struct GitRemote: Hashable {
     let name: String
     let pushLocation: String
     let fetchLocation: String
+    var branches: [GitBranch]?
```
It’s starting to feel like Go with all these optionals 😅

activcoding commented on CodeEdit/Features/SourceControl/Views/SourceControlStashView.swift:
```diff
@@ -0,0 +1,125 @@
+                    || sourceControlManager.switchToBranch != nil {
+                    if sourceControlManager.pullSheetIsPresented {
+                        try await sourceControlManager.pull(
+                            remote: sourceControlManager.operationRemote?.name ?? nil,
+                            branch: sourceControlManager.operationBranch?.name ?? nil,
```
The `?? nil` is redundant. Just use `sourceControlManager.operationBranch?.name`

activcoding reviewed (changes requested):
Almost ready to merge! I added one more comment. Once you address Khan's requested changes, we should be good to go. 🎉

austincondiff commented:
> It looks like theres a way to make the default selected branch in this menu be incorrect. This was after making a new branch, switching to it, and pressing the push button. It should default to the selected branch but it seems the operationBranch wasn't updated correctly from the repository outline view.

@thecoolwinter This seems to work as expected for me and selects the branch that I just switched to. Can you reproduce this with the latest?

austincondiff commented:
> I seem to be able to break the push mechanism on my machine:
> 
> For more details:
> 
> I created a branch, switched to it, pushed it to the remote, and switched back w/o closing the window
> After switching back, the push mechanism was broken.

Fixed this by reseting the operation variables when changing `pushSheetIsPresented` or  `pullSheetIsPresented`.

austincondiff commented:
@thecoolwinter and @activcoding I've address each of your issues you've raised. Please take a look and let me know if that works for you.

austincondiff commented:
> Could you provide more details on how you tested pulling changes? I can't figure how you did it with a local git repository.

Sure thing! In step number 7 under Testing in the PR description I say

> You can set up additional projects by repeating steps 1-5 so you can make changes elsewhere and test pull functionality.

So basically you can clone from the same local remote into another folder to have multiple projects open with the same remote. Then you can push from one and pull from the other.

activcoding commented on CodeEdit/Features/SourceControl/Views/SourceControlPullView.swift:
```diff
@@ -0,0 +1,90 @@
+        .frame(minWidth: 500)
+    }
+
+    func submit() {
```
Could you add documentation or change the function name? The function in SourceControlPushView has the same name and looks similar, which could lead to confusion

austincondiff replied:
You are submitting the form in both files. I thought that was clear enough. If the function were global I might understand.

activcoding replied:
Makes sense. Please add some documentation (a really short note is enough), like: ‘Pulls changes, shows stash sheet if needed

austincondiff replied:
Done

activcoding reviewed (changes requested):
Almost ready 👍🏻

austincondiff commented on CodeEdit/Features/NavigatorArea/SourceControlNavigator/Repository/Views/SourceControlNavigatorRepositoryView+contextMenu.swift:
```diff
@@ -23,20 +23,10 @@ extension SourceControlNavigatorRepositoryView {
     func contextMenu(for item: RepoOutlineGroupItem, branch: GitBranch) -> some View {
-        Button("Checkout") {
-            handleCheckout(branch)
+        Button("Switch...") {
```
It’s Switch in Xcode. It looks like git is moving this direction..
https://git-scm.com/docs/git-switch

activcoding reviewed (approved):
Ship it
